File: xendit_cards_session/example/integration_test/xendit_api_helper.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class XenditApiHelper {
  final String apiKey;
  final String baseUrl;
  
  XenditApiHelper({
    required this.apiKey,
    this.baseUrl = 'https://api.stg.tidnex.dev',
  });
  
  // Create a session for card payments
  Future<Map<String, dynamic>> createSession({
    required String referenceId,
    required int amount,
    required String locale,
    required String country,
    required String currency,
    required String customerId,
    required String sessionType,
    String mode = 'API',
    String? description,
    String? cardPaymentTokenId,
    Map<String, dynamic>? metadata,
  }) async {
    final url = Uri.parse('${this.baseUrl}/sessions');
    
    final headers = {
      'Authorization': apiKey.startsWith('Basic ') ? apiKey : 'Basic ${base64Encode(utf8.encode('$apiKey:'))}',
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };
    
    final body = jsonEncode({
      'mode': 'CARDS_SESSION_JS',
      'amount': amount,
      'locale': locale,
      'country': country,
      'currency': currency,
      'customer_id': customerId,
      'description': description,
      'reference_id': referenceId,
      'session_type': sessionType,
      'channel_properties': {
        'cards': {
          'skip_three_ds': true,
        },
      },
      'cards_session_js': {
        if (cardPaymentTokenId != null) 'card_payment_token_id': cardPaymentTokenId,
        'success_return_url': 'https://yourcompany.com/success',
        'failure_return_url': 'https://yourcompany.com/failure',
      },
      if (metadata != null) 'metadata': metadata,
    });
    
    print('Creating session at URL: $url');
    print('Request body: $body');
    
    try {
      final response = await http.post(
        url,
        headers: headers,
        body: body,
      ).timeout(
        const Duration(seconds: 10),
        onTimeout: () {
          print('HTTP request timed out after 10 seconds');
          throw Exception('Request timed out');
        },
      );
      
      print('Response status: ${response.statusCode}');
      print('Response body: ${response.body}');
      
      if (response.statusCode == 200 || response.statusCode == 201) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Failed to create session: ${response.statusCode} - ${response.body}');
      }
    } catch (e) {
      print('Error in HTTP request: $e');
      throw Exception('Error creating session: $e');
    }
  }
  
  // Extract session ID from session response
  String? extractSessionId(Map<String, dynamic> sessionResponse) {
    // The session ID is typically in the 'id' field
    return sessionResponse['payment_session_id'];
  }
  
  // Get session details
  Future<Map<String, dynamic>> getSession(String sessionId) async {
    final url = Uri.parse('${this.baseUrl}/sessions/$sessionId');
    
    final headers = {
      'Authorization': apiKey.startsWith('Basic ') ? apiKey : 'Basic ${base64Encode(utf8.encode('$apiKey:'))}',
      'Accept': 'application/json',
    };
    
    print('Getting session details from URL: $url');
    
    try {
      final response = await http.get(
        url,
        headers: headers,
      );
      
      print('Get session response status: ${response.statusCode}');
      print('Get session response body: ${response.body}');
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Failed to get session: ${response.statusCode} - ${response.body}');
      }
    } catch (e) {
      print('Error getting session: $e');
      throw Exception('Error getting session: $e');
    }
  }
}

File: androidApp/src/androidTest/java/com/cards/session/android/XenditApiHelper.kt
package com.cards.session.android

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

/**
 * Helper class for making API calls to Xendit for testing purposes
 */
class XenditApiHelper(
    private val apiKey: String,
    private val baseUrl: String
) {
    
    /**
     * Create a session for card payments
     */
    suspend fun createSession(
        referenceId: String,
        amount: Int,
        locale: String,
        country: String,
        currency: String,
        customerId: String,
        sessionType: String,
        description: String? = null,
        cardPaymentTokenId: String? = null,
        metadata: Map<String, Any>? = null
    ): Map<String, Any> = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/sessions")
        println("URL: ${url}");
        val connection = url.openConnection() as HttpURLConnection
        
        try {
            // Set up connection
            connection.requestMethod = "POST"
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("Accept", "application/json")
            
            // Set authorization header
            val authHeader = if (apiKey.startsWith("Basic ")) {
                apiKey
            } else {
                "Basic " + Base64.getEncoder().encodeToString("$apiKey:".toByteArray())
            }
            connection.setRequestProperty("Authorization", authHeader)
            
            // Create request body
            val requestBody = JSONObject().apply {
                put("mode", "CARDS_SESSION_JS")
                put("amount", amount)
                put("locale", locale)
                put("country", country)
                put("currency", currency)
                put("customer_id", customerId)
                put("reference_id", referenceId)
                put("session_type", sessionType)
                description?.let { put("description", it) }
                
                // Channel properties
                put("channel_properties", JSONObject().apply {
                    put("cards", JSONObject().apply {
                        put("skip_three_ds", true)
                    })
                })
                
                // Cards session JS properties
                put("cards_session_js", JSONObject().apply {
                    cardPaymentTokenId?.let { put("card_payment_token_id", it) }
                    put("success_return_url", "https://yourcompany.com/success")
                    put("failure_return_url", "https://yourcompany.com/failure")
                })
                
                metadata?.let { 
                    put("metadata", JSONObject(it))
                }
            }
            
            // Send request
            connection.outputStream.use { os ->
                os.write(requestBody.toString().toByteArray())
            }
            
            // Read response
            val responseCode = connection.responseCode
            val responseText = if (responseCode in 200..299) {
                connection.inputStream.bufferedReader().use { it.readText() }
            } else {
                connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
            }
            
            if (responseCode in 200..299) {
                // Parse JSON response to Map
                val jsonResponse = JSONObject(responseText)
                jsonResponse.toMap()
            } else {
                throw Exception("Failed to create session: $responseCode - $responseText")
            }
        } finally {
            connection.disconnect()
        }
    }
    
    /**
     * Extract session ID from session response
     */
    fun extractSessionId(sessionResponse: Map<String, Any>): String? {
        return sessionResponse["payment_session_id"] as? String
    }
    
    /**
     * Get session details
     */
    suspend fun getSession(sessionId: String): Map<String, Any> = withContext(Dispatchers.IO) {
        val url = URL("$baseUrl/sessions/$sessionId")
        val connection = url.openConnection() as HttpURLConnection
        
        try {
            // Set up connection
            connection.requestMethod = "GET"
            connection.setRequestProperty("Accept", "application/json")
            
            // Set authorization header
            val authHeader = if (apiKey.startsWith("Basic ")) {
                apiKey
            } else {
                "Basic " + Base64.getEncoder().encodeToString("$apiKey:".toByteArray())
            }
            connection.setRequestProperty("Authorization", authHeader)
            
            // Read response
            val responseCode = connection.responseCode
            val responseText = if (responseCode == 200) {
                connection.inputStream.bufferedReader().use { it.readText() }
            } else {
                connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
            }
            
            if (responseCode == 200) {
                // Parse JSON response to Map
                val jsonResponse = JSONObject(responseText)
                jsonResponse.toMap()
            } else {
                throw Exception("Failed to get session: $responseCode - $responseText")
            }
        } finally {
            connection.disconnect()
        }
    }
    
    /**
     * Extension function to convert JSONObject to Map
     */
    private fun JSONObject.toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        keys().forEach { key ->
            val value = get(key)
            map[key] = when (value) {
                is JSONObject -> value.toMap()
                else -> value
            }
        }
        return map
    }
}